spec/Candid.md

```diff
@@ -882,201 +882,150 @@ service { <name> : <functype>; <methtype>;* } <: service { <name> : <functype'>;
 
 ### Coercion
 
-This subtyping is implemented during the deserialisation of Candid at an expected type. As described in [Section Deserialisation](#deserialisation), the binary value is conceptually first _decoded_ into an abstract value, and then _coerced_ to the expected type.
+This subtyping is implemented during the deserialisation of Candid at an expected type. As described in [Section Deserialisation](#deserialisation), the binary value is conceptually first _decoded_ into the actual type and a value of that type, and then that value is _coerced_ into the expected type.
 
-In the following, we use `V` as an abstract data model of both the message in transit (`V`), as well as the result of the coercion (`V'`).  We re-use the syntax of the textual representation.
-
-Since messages are self-describing, i.e. carry type description, the rules below can refer to the given type of the message using the `<annval>` syntax (i.e. `(v : t)`) if the type matters. This is used in particular
- * To resolve overloading in number literals (`<primval>`)
- * To express the canonical value of type `reserved` as `(null : reserved)`.
- * To do coercion-time subtyping checks in the rule for `opt`
-
-This section describes the coercion as a ternary relation `V ~> V' : T` to describe that a value `V` can be coerced to a value `V'` of type `T`. The fields `V` and `T` can be understood as inputs and `V'` as the output of this relation. This relation is defined only when the type of `V` is a subtype of `T`.
+To model this, we defined, for every `t1, t2` with `t1 <: t2`, the function `c[t1<:t2] : t1 -> t2`. This function maps values of type `t1` to values of type `t2`, and is indeed total.
 
+to describe these values, we re-use the syntax of the textual representation, and use the the `<annval>` syntax (i.e. `(v : t)`) if necessary to resolve overloading.
 
 #### Primitive Types
 
-Values of primitive types coerce successfully at their own type:
+On primitve types, coercion is the identity:
 ```
-------------------------------------
-(<x> : <numtype>) ~> <x> : <numtype>
-
--------------------
-true ~> true : bool
-
----------------------
-false ~> false : bool
-
------------------------
-<text> ~> <text> : text
-
--------------------
-null ~> null : null
+c[<t> <: <t>](x) = x    for every <t> ∈ <numtype>, bool, text, null
```

Comment:
For hygiene, I think it's safer to expand the def for each case, so that you can limit the argument value to the right shapes as well.